util/election2/mysql/election.go

```diff
@@ -0,0 +1,280 @@
+// Copyright 2023 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mysql provides an implementation of leader election based on a SQL database.
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/trillian/util/election2"
+	"k8s.io/klog/v2"
+)
+
+type leaderData struct {
+	currentLeader string
+	timestamp     time.Time
+}
+
+// Election is an implementation of election2.Election based on a SQL database.
+type Election struct {
+	db         *sql.DB
+	instanceID string
+	resourceID string
+
+	currentLeader leaderData
+	leaderLock    sync.Cond
+
+	// If a channel is supplied with the cancel, it will be signalled when the election routine has exited.
+	cancel           chan *chan error
+	electionInterval time.Duration
+}
+
+var _ election2.Election = (*Election)(nil)
+
+// Await implements election2.Election
+func (e *Election) Await(ctx context.Context) error {
+	e.leaderLock.L.Lock()
+	defer e.leaderLock.L.Unlock()
+	if e.cancel == nil {
+		e.cancel = make(chan *chan error)
+		go e.becomeLeaderLoop(context.Background(), e.cancel)
+	}
+	if e.currentLeader.currentLeader == e.instanceID {
+		return nil
+	}
+	for e.currentLeader.currentLeader != e.instanceID {
+		e.leaderLock.Wait()
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			klog.Infof("Waiting for leadership, %s is the leader at %s", e.currentLeader.currentLeader, e.currentLeader.timestamp)
+		}
+	}
+	klog.Infof("%s became leader for %s at %s", e.instanceID, e.resourceID, e.currentLeader.timestamp)
+	return nil
+}
+
+// Close implements election2.Election
+func (e *Election) Close(ctx context.Context) error {
```

Comment:
Given this is essentially a wrapper around Resign, this election object could be re-used for a new election, which is slightly at odds with the interface definition. In practice, I don't think anything relies on this behaviour, and there aren't any resources that can be released, so I think it's safe. Does that match your understanding?

Reply:
Yes.

Do you want me to make the object unusable, to catch errors that might occur by calling other methods after `Close`?

Reply:
If you can think of a way of an elegant way of doing this, sure thing. I don't think it's super super important, it's just some extra care to make sure it's aligned with the interface definition :)